File: visitor-comparison/src/main/java/com/edwyn/demo/visitor/java25/review/OrderReviewService.java
package com.edwyn.demo.visitor.java25.review;

import com.edwyn.demo.visitor.java25.order.Discount;
import com.edwyn.demo.visitor.java25.order.DigitalProduct;
import com.edwyn.demo.visitor.java25.order.Order;
import com.edwyn.demo.visitor.java25.order.PhysicalProduct;
import com.edwyn.demo.visitor.java25.order.ServiceLine;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public final class OrderReviewService {
    public OrderReview review(Order order) {
        var invoiceLines = new ArrayList<String>();
        var issues = new ArrayList<String>();
        var subtotalCents = 0;
        var discountCents = 0;

        for (var item : order.items()) {
            invoiceLines.add(switch (item) {
                case PhysicalProduct product -> product.quantity()
                    + " x "
                    + product.name()
                    + " = "
                    + money(product.unitPriceCents() * product.quantity());
                case DigitalProduct product -> product.licenseCount()
                    + " licenses "
                    + product.name()
                    + " = "
                    + money(product.unitPriceCents() * product.licenseCount());
                case ServiceLine service -> service.name()
                    + " ("
                    + service.hours()
                    + "h) = "
                    + money(service.hourlyRateCents() * service.hours());
                case Discount discount -> "Discount "
                    + discount.code()
                    + " = "
                    + money(-discount.amountCents());
            });

            issues.addAll(switch (item) {
                case PhysicalProduct product -> {
                    var lineIssues = new ArrayList<String>();
                    var label = label(product.name(), "Physical product");
                    requireName(product.name(), "Physical product", lineIssues);
                    requirePositive(product.unitPriceCents(), label + " unit price", lineIssues);
                    requirePositive(product.quantity(), label + " quantity", lineIssues);
                    requirePositive(product.weightGrams(), label + " weight", lineIssues);
                    yield List.copyOf(lineIssues);
                }
                case DigitalProduct product -> {
                    var lineIssues = new ArrayList<String>();
                    var label = label(product.name(), "Digital product");
                    requireName(product.name(), "Digital product", lineIssues);
                    requirePositive(product.unitPriceCents(), label + " unit price", lineIssues);
                    requirePositive(product.licenseCount(), label + " license count", lineIssues);
                    yield List.copyOf(lineIssues);
                }
                case ServiceLine service -> {
                    var lineIssues = new ArrayList<String>();
                    var label = label(service.name(), "Service");
                    requireName(service.name(), "Service", lineIssues);
                    requirePositive(service.hourlyRateCents(), label + " hourly rate", lineIssues);
                    requirePositive(service.hours(), label + " hours", lineIssues);
                    if (service.hours() > 40) {
                        lineIssues.add(label + " cannot exceed 40 hours");
                    }
                    yield List.copyOf(lineIssues);
                }
                case Discount discount -> {
                    var lineIssues = new ArrayList<String>();
                    requireName(discount.code(), "Discount code", lineIssues);
                    requirePositive(discount.amountCents(), discount.code() + " discount amount", lineIssues);
                    yield List.copyOf(lineIssues);
                }
            });

            switch (item) {
                case Discount discount -> discountCents += discount.amountCents();
                case PhysicalProduct product -> subtotalCents += product.unitPriceCents() * product.quantity();
                case DigitalProduct product -> subtotalCents += product.unitPriceCents() * product.licenseCount();
                case ServiceLine service -> subtotalCents += service.hourlyRateCents() * service.hours();
            }
        }

        if (discountCents > subtotalCents) {
            issues.add("Discounts cannot exceed subtotal");
        }

        return new OrderReview(
            subtotalCents,
            discountCents,
            subtotalCents - discountCents,
            invoiceLines,
            issues
        );
    }

    private static String money(int cents) {
        return String.format(Locale.ROOT, "%.2f EUR", cents / 100.0);
    }

    private static String label(String value, String fallback) {
        return value == null || value.isBlank() ? fallback : value;
    }

    private static void requireName(String value, String label, List<String> issues) {
        if (value == null || value.isBlank()) {
            issues.add(label + " name is required");
        }
    }

    private static void requirePositive(int value, String label, List<String> issues) {
        if (value <= 0) {
            issues.add(label + " must be positive");
        }
    }
}

File: visitor-comparison/src/main/kotlin/com/edwyn/demo/visitor/kotlin/review/OrderReviewService.kt
package com.edwyn.demo.visitor.kotlin.review

import com.edwyn.demo.visitor.kotlin.order.DigitalProduct
import com.edwyn.demo.visitor.kotlin.order.Discount
import com.edwyn.demo.visitor.kotlin.order.Order
import com.edwyn.demo.visitor.kotlin.order.PhysicalProduct
import com.edwyn.demo.visitor.kotlin.order.ServiceLine
import java.util.Locale

class OrderReviewService {
    fun review(order: Order): OrderReview {
        val invoiceLines = mutableListOf<String>()
        val issues = mutableListOf<String>()
        var subtotalCents = 0
        var discountCents = 0

        order.items.forEach { item ->
            invoiceLines += when (item) {
                is PhysicalProduct ->
                    "${item.quantity} x ${item.name} = ${formatMoney(item.unitPriceCents * item.quantity)}"

                is DigitalProduct ->
                    "${item.licenseCount} licenses ${item.name} = ${formatMoney(item.unitPriceCents * item.licenseCount)}"

                is ServiceLine ->
                    "${item.name} (${item.hours}h) = ${formatMoney(item.hourlyRateCents * item.hours)}"

                is Discount ->
                    "Discount ${item.code} = ${formatMoney(-item.amountCents)}"
            }

            issues += when (item) {
                is PhysicalProduct -> buildList {
                    val label = item.name.ifBlank { "Physical product" }
                    validateName(item.name, "Physical product")
                    validatePositive(item.unitPriceCents, "$label unit price")
                    validatePositive(item.quantity, "$label quantity")
                    validatePositive(item.weightGrams, "$label weight")
                }

                is DigitalProduct -> buildList {
                    val label = item.name.ifBlank { "Digital product" }
                    validateName(item.name, "Digital product")
                    validatePositive(item.unitPriceCents, "$label unit price")
                    validatePositive(item.licenseCount, "$label license count")
                }

                is ServiceLine -> buildList {
                    val label = item.name.ifBlank { "Service" }
                    validateName(item.name, "Service")
                    validatePositive(item.hourlyRateCents, "$label hourly rate")
                    validatePositive(item.hours, "$label hours")
                    if (item.hours > 40) {
                        add("$label cannot exceed 40 hours")
                    }
                }

                is Discount -> buildList {
                    validateName(item.code, "Discount code")
                    validatePositive(item.amountCents, "${item.code} discount amount")
                }
            }

            when (item) {
                is Discount -> discountCents += item.amountCents
                is PhysicalProduct -> subtotalCents += item.unitPriceCents * item.quantity
                is DigitalProduct -> subtotalCents += item.unitPriceCents * item.licenseCount
                is ServiceLine -> subtotalCents += item.hourlyRateCents * item.hours
            }
        }

        if (discountCents > subtotalCents) {
            issues += "Discounts cannot exceed subtotal"
        }

        return OrderReview(
            subtotalCents = subtotalCents,
            discountCents = discountCents,
            totalCents = subtotalCents - discountCents,
            invoiceLines = invoiceLines,
            issues = issues,
        )
    }

    private fun formatMoney(cents: Int): String =
        String.format(Locale.ROOT, "%.2f EUR", cents / 100.0)

    private fun MutableList<String>.validateName(value: String, label: String) {
        if (value.isBlank()) {
            add("$label name is required")
        }
    }

    private fun MutableList<String>.validatePositive(value: Int, label: String) {
        if (value <= 0) {
            add("$label must be positive")
        }
    }
}
